insights: Add docs for capture group insights! (#29218)
sourcegraph/sourcegraph: doc/code_insights/explanations/automatically_generated_data_series.md

> # Automatically generated data series 

Reviewer: ```suggestion
# Detect and Track Patterns (Automatically generated data series)
```

Reviewer: Given that we've decided to use names for types of insights, docs should refer to these names as well.

Author: @AlicjaSuska good point, thanks. Modifying slightly
```suggestion
# Detect and track patterns with automatically generated data series 
```

> > Note: in Sourcegraph 3.35 this feature is in its earliest version. Stability and additional features coming in following releases. 

Reviewer: ```suggestion
> Note: "Detect and track patterns" insights are available in Sourcegraph 3.35 and above. In Sourcegraph 3.35 this feature is in its earliest version. Stability and additional features coming in the following releases. 
```

Reviewer: I suggest keeping all the information about the version, availability and stability in this note. It means removing the 'On Sourcegraph 3.35..' from the first paragraph.

> On Sourcegraph 3.35 or later, Code Insights with automatically generated data series allow you to detect and track versions and patterns within your codebase. You can use these insights to track versions of languages, packages, terraform, docker images, or anything else that can be captured with a [regular expression capture group](#regular-expression-capture-group-resources).

Reviewer: ```suggestion
Code Insights with automatically generated data series ("Detect and Track Patterns") allow you to detect and track versions and patterns within your codebase. You can use these insights to track versions of languages, packages, terraform, docker images, or anything else that can be captured with a [regular expression capture group](#regular-expression-capture-group-resources).
```

Reviewer: We also use the link to 'track insight' info in the creation UI when explaining limitation about the all repos insight ([Figma](https://www.figma.com/file/qomcHxvHjrwfS3We6NVHGh/Automatically-generated-data-series-%5Breview%5D?node-id=1809%3A31518))

Author: Thanks @AlicjaSuska ! I will add those docs in a separate PR, after finishing this one first.

Author: ```suggestion
Code Insights with automatically generated data series allow you to "detect and track patterns" and versions within your codebase. You can use these insights to track versions of languages, packages, terraform, docker images, or anything else that can be captured with a [regular expression capture group](#regular-expression-capture-group-resources).
```

@AlicjaSuska I'm not adding "Detect and Track Patterns" to this sentence twice since the words were already there

Reviewer: Apologies, maybe I haven't explained the objective here - my goal was to explicitly call out the insight type name we use in the selection screen, even if words are already used in the sentence. Overall, I think that this connection to the insight type should be stronger in the documentation to help people create a mental connection.

## Automatic generation of individual series

Match patterns can be returned with a regular expression containing a capture group, or a pattern within parentheses. 

For example, `file:\.pom$ <java\.version>(.*)</java\.version>` will match all of: 

```
<java.version>1.5</java.version>
<java.version>1.7</java.version>
<java.version>1.8</java.version>
```

> Code Insights will find all matches, and then automatically generate a data series and series color for each unique value of the capture group. In this case, the chart would show data series for 1.5, 1.7, and 1.8, with the values being the number of matches for each unique value. 

Reviewer: ```suggestion
Code Insights will find all matches, and then automatically generate a data series and assign color for each unique value of the capture group. In this case, the chart would show data series for 1.5, 1.7, and 1.8, with the values being the number of matches for each unique value. 
```

Author: > 

```suggestion
Code Insights will find all matches, and then automatically generate a data series and color for each unique value of the capture group. In this case, the chart would show data series for 1.5, 1.7, and 1.8, with the values being the number of matches for each unique value. 
```

> ## Automatic updates

Reviewer: ```suggestion
## New data series automatically added to your insight
```

'Automatic updates' can be slightly misleading as a header, especially given that people need to update Sourcegraph so it's probably going to be their first association. I'm not set on new wording I've proposed, it can be improved for sure but I think it shows the direction.

Author: ```suggestion
## New matching data gets automatically added 
```
Thanks @AlicjaSuska ! Good point. Tweaked wording slightly

> Capture groups will automatically show new matches when they appear in your codebase; you do not need to update or manually re-create the insights. 

Reviewer: ```suggestion
Capture groups will automatically create new data series for new matches as they appear in your codebase. You do not need to update or manually re-create the insights to track newly added versions or patterns.
```

I'd like to support the last sentence with a clear use case to answer the question 'why would I want to update the insight manually in the first place?'

Author: Makes sense! Thanks.
